library/std/src/sys/pal/unix/futex.rs

```diff
@@ -135,6 +139,186 @@ pub fn futex_wake_all(futex: &AtomicU32) {
+        let timeout_ms = timeout.as_micros().clamp(1, u32::MAX as u128) as u32;
+
+        if let Some(wait) = os_sync_wait_on_address_with_timeout.get() {
+            let r = unsafe {
+                wait(
+                    addr,
+                    value,
+                    size,
+                    OS_SYNC_WAIT_ON_ADDRESS_NONE,
+                    OS_CLOCK_MACH_ABSOLUTE_TIME,
+                    timeout_ns,
+                )
+            };
+
+            // We promote spurious wakeups (reported as EINTR) to normal ones for
+            // simplicity and because Apple's documentation is unclear as to what the
+            // unit for the deadline of `os_sync_wait_on_address_with_deadline` is,
+            // making it hard to support timeouts in a fashion similar to the Linux
+            // futex implementation.
+            r != -1 || super::os::errno() != libc::ETIMEDOUT
+        } else if let Some(wait) = __ulock_wait2.get() {
+            unsafe {
+                wait(UL_COMPARE_AND_WAIT | ULF_NO_ERRNO, addr, value, timeout_ns, 0)
+                    != -libc::ETIMEDOUT
+            }
+        } else if let Some(wait) = __ulock_wait.get() {
+            unsafe {
+                wait(UL_COMPARE_AND_WAIT | ULF_NO_ERRNO, addr, value, timeout_ms)
+                    != -libc::ETIMEDOUT
+            }
+        } else {
+            panic!("your system is below the minimum supported version of Rust");
+        }
+    } else {
+        if let Some(wait) = os_sync_wait_on_address.get() {
+            unsafe {
+                wait(addr, value, size, OS_SYNC_WAIT_ON_ADDRESS_NONE);
+            }
+        } else if let Some(wait) = __ulock_wait.get() {
+            unsafe {
+                wait(UL_COMPARE_AND_WAIT | ULF_NO_ERRNO, addr, value, 0);
+            }
+        } else {
+            panic!("your system is below the minimum supported version of Rust");
+        }
+        true
+    }
+}
+
+#[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos", target_os = "watchos",))]
+pub fn futex_wake(futex: &AtomicU32) -> bool {
+    use crate::io::Error;
+    use crate::mem::size_of;
+    use apple::*;
+
+    let addr = futex.as_ptr().cast();
+    if let Some(wake) = os_sync_wake_by_address_any.get() {
+        unsafe { wake(addr, size_of::<u32>(), OS_SYNC_WAKE_BY_ADDRESS_NONE) == 0 }
+    } else if let Some(wake) = __ulock_wake.get() {
+        // __ulock_wake can get interrupted, so retry until either waking up a
+        // waiter or failing because there are no waiters (ENOENT).
+        loop {
+            let r = unsafe { wake(UL_COMPARE_AND_WAIT | ULF_NO_ERRNO, addr, 0) };
+
+            if r >= 0 {
+                return true;
+            } else {
+                match -r {
+                    libc::ENOENT => return false,
+                    libc::EINTR => continue,
+                    err => panic!("__ulock_wake failed: {}", Error::from_raw_os_error(err)),
+                }
+            }
+        }
```

Comment:
Hmm, are you sure this is correct? [The implementation of `os_sync_wake_by_address_any`](https://github.com/apple-oss-distributions/libplatform/blob/a00a4cc36da2110578bcf3b8eeeeb93dcc7f4e11/src/os/lock.c#L1468-L1506) seems to just call `__ulock_wake` without a loop? So either we need to loop both places, or in none of them?

Same goes for `futex_wake`.